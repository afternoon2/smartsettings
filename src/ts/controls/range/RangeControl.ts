import { Control } from '../control/Control';
import {
  RangeControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/range.sass';
import ControlStyles from '../../../sass/control.sass';

export class RangeControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="range"
      class="${Styles.range}"
      ${state.value && typeof state.value === 'number' ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly' : ''}
      ${state.disabled ? 'disabled' : ''}
    />
  `;

  constructor(props: RangeControlProps) {
    super(props, RangeControl.template);
    this.controlElement = this.element.querySelector('[type="range"]') as HTMLInputElement;
    this.updateMain();
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
    (this.element.querySelector(`.${Styles.rangeValue}`) as HTMLSpanElement).textContent = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }

  private updateMain = () => {
    const main = this.element.querySelector(`.${ControlStyles.control__content}`) as HTMLElement;
    main.classList.add(`${ControlStyles['control__content--range']}`);
    main.insertAdjacentHTML('beforeend', `<span class="${Styles.rangeValue}">
      ${typeof this.state.value === 'number' ? this.state.value : ''}
    </span>`);
  }
}