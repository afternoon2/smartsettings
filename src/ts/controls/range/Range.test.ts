import { RangeControl } from './RangeControl';

let control: RangeControl;

const listener = jest.fn();

describe('Range control', () => {
  beforeEach(() => {
    control = new RangeControl({
      id: 'id',
      name: 'name',
      options: {
        value: 0,
        min: 0,
        max: 10,
        step: 0.1,
        disabled: false,
      },
      parentElement: document.body,
      listener,
    });
  });

  test('If it matches snapshot after initial load', () => {
    expect(control.element).toMatchSnapshot();
  });

  test('Disable', () => {
    control.disable();
    expect(control.element).toMatchSnapshot();
  });

  test('Enable', () => {
    control.disable();
    control.enable();
    expect(control.controlElement.hasAttribute('disabled')).toBe(false);
    expect(control.element).toMatchSnapshot();
  });

  test('Get and set value', () => {
    control.value = 4;
    expect(control.value).toBe(4);
  });

  test('If it assings user listener correctly', () => {
    expect(listener).toHaveBeenCalledTimes(4);
  });
});