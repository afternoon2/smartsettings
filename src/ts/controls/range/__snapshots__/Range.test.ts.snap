// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range control Disable 1`] = `
<div
  class="control disabled"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content control__content--range"
    >
      <input
        class="range"
        disabled="true"
        max="10"
        min="0"
        step="0.1"
        type="range"
      />
      
  
      <span
        class="rangeValue"
      >
        
      0
    
      </span>
    </div>
  </main>
</div>
`;

exports[`Range control Enable 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content control__content--range"
    >
      <input
        class="range"
        max="10"
        min="0"
        step="0.1"
        type="range"
      />
      
  
      <span
        class="rangeValue"
      >
        
      0
    
      </span>
    </div>
  </main>
</div>
`;

exports[`Range control If it matches snapshot after initial load 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content control__content--range"
    >
      <input
        class="range"
        max="10"
        min="0"
        step="0.1"
        type="range"
      />
      
  
      <span
        class="rangeValue"
      >
        
      0
    
      </span>
    </div>
  </main>
</div>
`;
